### DISCUSSION TITLE
Dynamically declare `*Target` properties with TS

### DISCUSSION THREAD

**Discussion by mlwyatt** [stimulusjs/stimulus#831] [General]
I've looked over #121 and #466 and while they are similar, they aren't quite what I'm looking for. I know when using TS I need to `declare` the `*Target` properties I want to use. I'm fine with that but I'm wondering if that can done dynamically.

`declare`ing all of the properties isn't too bad when there's only 1 or 2 `targets`, but sometimes I have numerous `targets` and would like to avoid `declare`ing every one of them when they're all very similar. It's a similar situation for `values` and `outlets` but I use those far less than `targets`

Is there a way to dynamically `declare` these properties or to tell TS that `Controller` does correctly implement `ControllerWithTarget<Targets>` below (while avoiding `@ts-ignore` and similar) ?

```typescript
import { Controller } from '@hotwired/stimulus';

namespace Transform {
  export type HasTarget<T> = {
    [K in keyof T as `has${Capitalize<string & K>}Target`]: boolean;
  };
  export type Target<T> = {
    [K in keyof T as `${string & K}Target`]: T[K];
  };
  export type Targets<T> = {
    [K in keyof T as `${string & K}Targets`]: Array<T[K]>;
  };
}

const targets = [
  'button',
  'container',
] as const;

type TargetList = typeof targets;

interface Targets extends Record<TargetList[number], HTMLElement> {
  button: HTMLButtonElement;
  container: HTMLDivElement;
}

type ControllerWithTarget<T> =
  & Transform.HasTarget<T>
  & Transform.Target<T>
  & Transform.Targets<T>;

// Trying to achieve something like this
export default class extends Controller implements ControllerWithTarget<Targets> {
  static targets = [...targets];

  // declare readonly hasButtonTarget: boolean;
  // declare readonly buttonTarget: HTMLButtonElement;
  // declare readonly buttonTargets: Array<HTMLButtonElement>;

  // declare readonly hasContainerTarget: boolean;
  // declare readonly containerTarget: HTMLDivElement;
  // declare readonly containerTargets: Array<HTMLDivElement>;

  disableButton(): void {
    if ( this.hasButtonTarget ) {
      this.buttonTarget.disabled = true;
    }
  }
}
```

- [SO Question](https://stackoverflow.com/q/76900171/5909281)
- [TS playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgYQgOxlCAbbBTKOAXzgDMsQ4ByAAQAsIYB3YKPAEwHoBnGUAV2z9uVANwAocWgCGIPNzDSAxnjgAVKNLTdS0SgnFw4eAB6RYcGAE8wqgBLTua6VADmeGAB41APjgBeREMjOABtAGk4YDQ4AGs8KwhSdThHOAADOkcAEgRkaTBgGGlsYAAvPE9eKGjXOAAyOHCfImc3D3SAXQAuOAAjCBw8LQkjIlHjM2h4a1t1F3cvXwCgkLDI6LiEpJS09Nzq2oam1oWOnvUIzonx4NNzGZtVNsXubz9AgzWIqJj4xOSalS3AyB0wR0a4VO7Rg3C6vQAglBNFZvFcfDcJERJEp0LxLGdYStQsEqH1+DAYOgqAAaUm4jDSaIEWniTrAuAM3gScSzZ6EgAywHxgT5O2KMO4POiMAIpGU-Mlk1laHYIIASnhcVB2N5BcKYKE0PwQH0CJ0aXA7GoALICgCi+DkGD8X36FKpaF61rtACEPehHXhnTAJgzisyoN7bQKACLAABuQZDWMkfJQ6EwOHwUAA6kU6C8PO8AsFGhotDo9AA6BxOQnvMvqTTaXRQEDVotLHxNiutmtdt6+HniTicZtWI5U1JKOjAPAJ1TcCByGBztB1UrxSxz7jie7TODsPDywTwJTYRwg0wqtUZjBYXAEKLgJ14DAg1AP7MEfNrrt6pKrrBLw0h8EoBJKoEoTVrBEqvNckhGGOR5apebBwGw0jsOg2BWHAWTcP6lLoF2vQDEMIzBChx4Xi4qhYThaB4e6JFoGRVoxsRnrJu+obUeOtHoQxwxMSx5JsYOiLItIqI+gK3GBm+Lo8shglofRmGibh+GEV+EZoAQHEUfgVFqahdEYYxOmcpmTKGVAHHyfGSbKfx5lCZp1nMfh4b2UZhLcNJKKeM5ia8SpSFHsK0h9PgiloAAFAAlL0CYQMA7CrCEwDJIlO7CtWhEJV2cDJdlaxroVEmel21bsDFcUcCsmD8HgExjME2JEEAA) 

Any help is greatly appreciated, thanks!

**Comment by marcoroth** [comment]
Hey @mlwyatt, I know this might not be exactly what you are looking for, but I find the pattern proposed in the [`stimulus-decorators`](https://github.com/vytant/stimulus-decorators) package kinda elegant.

**Comment by mlwyatt** [comment]
> Hey @mlwyatt, I know this might not be exactly what you are looking for, but I find the pattern proposed in the [`stimulus-decorators`](https://github.com/vytant/stimulus-decorators) package kinda elegant.

Thanks for reaching out! I agree that it's elegant but it's also a bit outdated (no support for outlets) and not quite what I'm after. I think what I want just isn't possible with the current state of TS

**Comment by ajaishankar** [comment]
@mlwyatt see https://github.com/hotwired/stimulus/issues/724 for a possible way to implement this...

**Comment by ajaishankar** [comment]
@mlwyatt give this a go

https://github.com/ajaishankar/stimulus-typescript
